nncf/torch/search_blocks.py

```diff
@@ -0,0 +1,510 @@
+    if not allow_nested_blocks:
+        sorted_blocks = remove_nested_blocks(sorted_blocks)
+    building_blocks_in_orig_graph = restore_node_name_in_orig_graph(sorted_blocks, orig_graph)
+
+    return building_blocks_in_orig_graph
+
+
+def remove_nested_blocks(sorted_blocks):
+    return [list(group_block)[-1] for _, group_block in groupby(sorted_blocks, lambda block: block.start_node.main_id)]
+
+def get_group_of_dependent_blocks(blocks):
+    groups = {}
+    idx = 0
+    groups = { idx: [] }
+    for i in range(len(blocks) - 1):
+        start_node_key_i1 = blocks[i + 1][0]
+        end_node_key_i  = blocks[i][1]
+        if start_node_key_i1 == end_node_key_i:
+            groups[idx].append(i)
+        else:
+            groups[idx].append(i)
+            idx +=1
+            groups[idx] = []
+    groups[idx].append(len(blocks) - 1)
+
+    return groups
+
+def get_building_blocks_info(bblocks: List[BuildingBlock], compressed_model: NNCFNetwork) -> List[BuildingBlockInfo]:
+    bblocks_info = []
+    for block in bblocks:
+        op_addresses = get_all_node_op_addresses_in_block(compressed_model, block)
+        modules = get_all_modules_in_blocks(compressed_model, op_addresses)
+        block_type = get_type_building_block(op_addresses)
+        bblocks_info.append(BuildingBlockInfo(block, op_addresses, modules, block_type))
+    return bblocks_info
+
+def get_all_node_op_addresses_in_block(compressed_model: NNCFNetwork, block: BuildingBlock):
+    graph = compressed_model.get_original_graph()
+    nx_graph = graph.get_nx_graph_copy()
+    start_node, end_node = block
+    start_node_key, end_node_key = None, None
+    for node in nx_graph._node.values():
+        if start_node == str(node['node_name']):
+            start_node_key = node['key']
+        if end_node == str(node['node_name']):
+            end_node_key = node['key']
+    simple_paths = nx.all_simple_paths(nx_graph, start_node_key, end_node_key)
+    op_adresses = set()
+    for node_keys_in_path in simple_paths:
+        for node_key in node_keys_in_path:
+            op_adresses.add(OperationAddress.from_str(nx_graph._node[node_key]['node_name']))
+    start_op_address = OperationAddress.from_str(nx_graph._node[start_node_key]['node_name'])
+    op_adresses.remove(start_op_address)
+    return op_adresses
+
+def get_all_modules_in_blocks(compressed_model: NNCFNetwork,
+                              op_adresses_in_blocks: List[OperationAddress]) -> List[torch.nn.Module]:
+    modules = []
+    for op_address in op_adresses_in_blocks:
+        if op_address.operator_name in NNCF_MODULES_OP_NAMES:
+            modules.append(compressed_model.get_module_by_scope(op_address.scope_in_model))
+    return modules
+
+def get_type_building_block(op_addresses_in_block: List[OperationAddress])-> BuildingBlockType:
+    count_matmul = 0
+    count_fc = 0
+    for op_address in op_addresses_in_block:
+        if op_address.operator_name == 'matmul':
```

Comment:
using MetaType instead of string would be better

Reply:
Done